### DISCUSSION TITLE
Allow hard returns in meeting notes field

### DISCUSSION THREAD

**Discussion by OSAwebsite** [code4recovery/12-step-meeting-list#1061] [General]
When I import a new meeting via .csv, I noticed that whatever is in the Notes field (the field below the checkboxes in the back end) is stripped of hard returns. Is there a way around this?

Thanks

**Comment by tim-rohrer** [comment]
Hi Nicky,

I just tested this on my dev site and it the new lines were returned. I don't have time to test further today, but if you could provide some screen caps of what you're seeing, that might help figure it out.

And what app are you using to modify the csv file?

**Comment by OSAwebsite** [comment]
Thanks for testing, Tim.
The csv is created in text edit. It is stripped of formatting.
It comes from a database and I am attaching a test.csv that shows that it has in fact returns.
I selected the whole thing and pasted it into Word to check that there were no line returns instead of real returns but they look like real returns to me. This also happens in the current interface.
I tried changing a single return into a double, to see if maybe that would solve the issue, but it does not seem to be.
Let me know if you need more information.

Thanks. [Nicky]([url](url))
[test4.csv](https://github.com/code4recovery/12-step-meeting-list/files/10470927/test4.csv)

**Reply by tim-rohrer** [reply]
I completed several rounds of testing, including using your test4 file. If I start with your file, import into TSML, add carriage returns in the notes, export it, edit it with TextEdit, save it, reimport it, the new lines remain and all is well. 

However, if I simply import your file, the extra lines are not present.

For next steps, I want to focus on your original file and its history. When you wrote "text edit", is that the name of the app you're using? I'm assuming you're on a Windows machine since you mentioned "Word". There are subtle differences in how Windows and Mac/Linux systems handle new lines.  That said, most of this is so well documented that the shortcomings arising from it have been resolved, but not always. I suspect that whatever process you're using is stripping the linefeeds, but not the carriage returns. 

For example, if I take your original file, and `cat` it in a terminal, I should see the column headings first, and then in the next line, the data. But I don't, instead, I'm only getting one line that looks like this:

```sh
Password: 878352","St. Andrew's Memorial Presbyterian Church","24 Stavebank Road, Mississauga, ON, L5G2T5, Canada","Mississauga","Lakeshore Rd and Stavebank - Front door - upstairs front room.","Suburban West Port Credit AFG(D16)","District 16","Group ID: 16469 - Program: Al-Anon - Region: Mississauga, Oakville, Port Credit","Families and Friends Only,Online Meeting,Location Temporarily Closed,Concurrent with Alateen Meeting","https://zoom.us/j/81387697701",""% 
```

What database is it coming from? How is it pulled from the database? What are you doing with the data after that, but before you try importing it into TSML?

**Comment by OSAwebsite** [comment]
Hi Tim,

I am in fact on a MacIntosh. The program I use to prepare the import is TextEdit, a very simple little text editor, which opens my export file as text only and I then save it in the .csv format.  The word processor Microsoft Word has a Mac version as well and that was the program I was referring to: it is easy to make all characters visible in Word; not so much in TextEdit. I just wanted to make sure that what I see as hard return is in fact a hard return and it seems to me that it is/they are.

It is strange that you do not see the column headings because I do. I created a new file, to make sure I am not missing something. Then content of the file looks like this: (without the === dividers, those are just to make it clear below where the file content begins and ends)

===========
Time,Day,Name,Notes,Location,Address,Region,Location Notes,Group,District,Group Notes,Types,Conference_URL,Conference_phone
"8 PM","Sunday","(online) Faith, Hope, & Serenity AFG - Orillia (D1)","Group ID: 46195

Meeting focus: Discussion meeting

Speaker meeting last Sunday of any month that contains 5 Sundays

Access directly with the 'Join with Zoom' button on the meetings page of the Ontario website. Or enter Zoom details manually.

Meeting ID: 84450540289

Password: 19404269","St. Athanasius Anglican Church","10 Westmount Dr N, Orillia, ON, L3V 6B8, Canada","Orillia","Wheelchair Accessible.","Faith, Hope, & Serenity AFG - Orillia (D1)","District 1","Group ID: 46195 - Program: Al-Anon - Region: Angus, Barrie, Bracebridge, Huntsville, Innisfil, Midland, Orillia, Parry Sound, Sprucedale, Wasaga BeachBorden
Huntsville
Burk's Falls
Bracebridge
Wasaga Beach
Midland
Barrie
Orillia
Parry Sound
","Families Friends and Observers Welcome","https://zoom.us/j/84450540289",""

==========

File is attached.

When I paste the content in Word, it looks like the attached picture file and you can see that the hard returns are present.

I just re-imported in my staging site and the field is stripped of returns. Same in the live environment.

Because there is no preview of the field mapping (as, for instance, a plug-in like The Events Calendar offers), I cannot see what is wrong with the file. I get no errors.

**Import file:**
[import_test.csv](https://github.com/code4recovery/12-step-meeting-list/files/10472447/import_test.csv)

**Picture of hard returns as seen in Microsoft Word:**
![Screenshot 2023-01-21 at 10 28 36 AM](https://user-images.githubusercontent.com/59859095/213874210-1f7f0b1f-c2c1-4b05-b29f-19bdbd6d9cb6.png)

The export file is created from one field per group in a program called Claris FileMaker. This is a cross platform suite of programs that works on Mac, Windows, etc.

**Reply by tim-rohrer** [reply]
After I wrote that I suspected you were on Windows, I realized you might have Word on Mac. The problem with using something like Word (I can do the same thing with Pages) is that we do not know truly what they mean by a simple paragraph marking at the end of the line. 

The good news is that you can repeat the test I did by opening a Terminal and typing `cat import_test.csv` and you'll see a result like this:"

```sh
➜  ~ cat ~/Desktop/import_test.csv 
Password: 19404269","St. Athanasius Anglican Church","10 Westmount Dr N, Orillia, ON, L3V 6B8, Canada","Orillia","Wheelchair Accessible.","Faith, Hope, & Serenity AFG - Orillia (D1)","District 1","Group ID: 46195 - Program: Al-Anon - Region: Angus, Barrie, Bracebridge, Huntsville, Innisfil, Midland,","Families Friends and Observers Welcome","https://zoom.us/j/84450540289",""%                                                                        ➜  ~ 
```
This is telling me that something isn't kosher in the encoding exported by FileMaker. 

Check out this thread, it may be related but I don't use (or have) FileMaker so I can't experiment.

https://community.claris.com/en/s/question/0D50H00006dsiRASAY/export-records-with-crlf-on-a-mac

I'm confident that TSML is working correctly. I suggest focusing on different methods to export data from FileMaker. You should be able to cat the export and see all of the data.

**Comment by OSAwebsite** [comment]
OK, I figured it out. 
When I inspected my own .csv original in Visual Studio Code it detected CRLF characters or what I describe as hard returns.
That is what I expected to see and had no reason to expect that those characters, as long as they exist within double quotes in the CSV, would translate in a line feed.

However, when I performed an export from my live website, Visual Studio Code detected no CRLF characters, not even in the entries that I had adjusted on the website, manually, after import. It contained just LF characters.

I have just adjust my own code in FileMaker, where the CSV is created with the column headers and the content, to use LF instead of CRLF characters and now the hard returns are preserved.

I never thought of checking the difference between the two csv outputs, but I am glad I finally found out how to fix it.

**Comment by OSAwebsite** [comment]
The Visual Studio Code program (free) will always show, in the footer, which characters are used in a document, even the ones that are invisible. I just never thought of comparing the two: an export from an existing listing and my .csv. VSC also exists for Windows, by the way, and it will do the same there. I am glad I got this solved. If I ever had to do a full database replacement on the webpage, I would have otherwise had to manually change all of the listings that do not have the returns.

I have left some notes in another discussion about the new TSML interface. I am not using it yet, because of the header and footer being identical in list mode and details mode. I kind of hope that this will be changed, so that I can customise both. For now, I will stay with the old interface.

Thanks for looking into this and bringing up the idea of comparing the two outputs. We learn new things every day and for that I am grateful!
